### PR TITLE
OV-52: add chat controller

### PR DESCRIPTION
**Pull request by sergiy4**
Added logic for storing chat history and an endpoint for interacting with the chat

Description:
- Implemented functionality for storing chat history.
-  Added logic for deleting old messages and clearing the chat history.

ROUTES:

1. POST http://localhost:3001/api/v1/chat/ ( for text generation )
2. DELETE http://localhost:3001/api/v1/chat/ ( for deleting the session )

Photo 1: Generating text using OpenAI

Photo 2: Providing context

Photo 3: Storing chat history (the chat has context because it knows my name)

image
Photo 4: Deleting the session

Photo 5: The chat doesn’t know my name because the session has been deleted

### CODE REVIEW

File: backend/src/bundles/chat/chat.controller.ts [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+                this.clearChat(
+                    options as ApiHandlerOptions<{
+                        session: FastifySessionObject;
+                    }>,
+                ),
+        });
+
+        this.addRoute({
+            path: ChatPath.ROOT,
+            method: HTTPMethod.POST,
```
Comment (Sanchousina) [inline backend/src/bundles/chat/chat.controller.ts:54]: Maybe I don't understand something, but why are you deleting the session with post and not delete method? It seems to me a little counterintuitive...

File: backend/src/bundles/chat/chat.controller.ts [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+    private async generateChatAnswer(
+        options: ApiHandlerOptions<{
+            body: GenerateTextRequestDto;
+            session: FastifySessionObject;
+        }>,
+    ): Promise<ApiHandlerResponse> {
+        const { body, session } = options;
+
+        this.openAIService.addMessageToHistory(
```
Comment (Sanchousina) [inline backend/src/bundles/chat/chat.controller.ts:94]: Why not handling the logic in separate chat service and not directly in chat controller? This concerns other controller methods too

File: backend/src/bundles/chat/chat.controller.ts [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    private openAIService: OpenAIService;
+
+    public constructor(logger: Logger, openAIService: OpenAIService) {
+        super(logger, ApiPath.CHAT);
+
+        this.openAIService = openAIService;
+
+        this.addRoute({
+            path: ChatPath.ROOT,
+            method: HTTPMethod.PUT,
```
Comment (Sanchousina) [inline backend/src/bundles/chat/chat.controller.ts:28]: Also why did you decide to use put method for generating chat answer? Isn't post method here more appropriate?

File: backend/src/bundles/chat/open-ai.service.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        chatHistory: Message[],
+        userMessage: string,
+        role: ValueOf<typeof OpenAIRole>,
+    ): void {
+        const newUserMessage = {
+            content: userMessage,
+            role,
+        };
+
+        chatHistory.push(newUserMessage);
```
Comment (Sanchousina) [inline backend/src/bundles/chat/open-ai.service.ts:41]: Isn't it better practice to keep parameters immutable? I see that you change messages array parameter in other methods as well.

https://github.com/ryanmcdermott/clean-code-javascript?tab=readme-ov-file#avoid-side-effects-part-2

File: backend/src/bundles/chat/char.service.ts
Comment (o-nedashkivska) [inline backend/src/bundles/chat/char.service.ts:1]: Please rename this file to chat.service.ts

File: backend/src/bundles/chat/chat.controller.ts [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+            handler: (options) =>
+                this.deleteSession(
+                    options as ApiHandlerOptions<{
+                        session: FastifySessionObject;
+                    }>,
+                ),
+        });
+    }
+
+    private async generateChatAnswer(
```
Comment (o-nedashkivska) [inline backend/src/bundles/chat/chat.controller.ts:74]: Please, add swagger docs for all handlers

**Review by o-nedashkivska — DISMISSED** [review]

**Review by o-nedashkivska — DISMISSED** [review]

File: backend/src/bundles/chat/chat.service.ts [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        chatHistory: Message[],
+        userMessage: string,
+        role: ValueOf<typeof OpenAIRole>,
+    ): void {
+        const newUserMessage = {
+            content: userMessage,
+            role,
+        };
+
+        chatHistory.push(newUserMessage);
```
Comment (Sanchousina) [inline backend/src/bundles/chat/chat.service.ts:29]: You return new messages array in deleteOldMessages method, but you still mutate chatHistory parameter here

File: backend/src/bundles/chat/chat.controller.ts [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+
+        return {
+            payload: { generatedText },
+            status: HttpCode.OK,
+        };
+    }
+
+    /**
+     * @swagger
+     * /chat/clear:
```
Comment (nikita-remeslov) [inline backend/src/bundles/chat/chat.controller.ts:138]: chat/:id will be good here

File: backend/src/bundles/chat/chat.controller.ts [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+     *          description: Failed operation
+     *          content:
+     *            application/json:
+     *              schema:
+     *                  type: object
+     *                  properties:
+     *                    isDeleted:
+     *                      type: boolean
+     */
+    private deleteSession(
```
Comment (nikita-remeslov) [inline backend/src/bundles/chat/chat.controller.ts:196]: why do we need 2 methods clear and delete?

**Review by nikita-remeslov — CHANGES_REQUESTED** [review]

File: backend/src/bundles/chat/chat.service.ts (hunk 6 above)
Comment (sergiy4) [reply backend/src/bundles/chat/chat.service.ts]: Done

**Review by Sanchousina — DISMISSED** [review]

File: backend/.env.example [hunk 9]
```diff
@@ -20,3 +20,13 @@ AWS_SECRET_ACCESS_KEY=see-in-slack
 AWS_S3_REGION=eu-north-1
 AWS_S3_BUCKET_NAME=bsa-2024-outreachvids
 AWS_CLOUDFRONT_DOMAIN_ID=d2tm5q3cg1nlwf
+
+#
+# OPEN AI
+#
+OPEN_AI_KEY=''
```
Comment (nikita-remeslov) [inline backend/.env.example:27]:
```suggestion
OPEN_AI_KEY=SOME_SECRET_KEY
```

**Review by nikita-remeslov — DISMISSED** [review]

**Review by nikita-remeslov — APPROVED** [review]

File: backend/src/bundles/chat/chat.controller.ts [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+import { ApiPath } from '~/common/enums/enums.js';
+import { HttpCode, HTTPMethod } from '~/common/http/http.js';
+import { type Logger } from '~/common/logger/logger.js';
+import { MAX_TOKEN } from '~/common/services/open-ai/libs/constants/constants.js';
+
+import {
+    ChatPath,
+    OpenAIRole,
+} from '../../common/services/open-ai/libs/enums/enums.js';
+import { type OpenAIService } from '../../common/services/open-ai/open-ai.service.js';
```
Comment (sofiia-trokhymchuk) [inline backend/src/bundles/chat/chat.controller.ts:17]: Use relative imports

**Review by sofiia-trokhymchuk — CHANGES_REQUESTED** [review]

**Review by Sanchousina — DISMISSED** [review]

**Review by o-nedashkivska — DISMISSED** [review]

**Review by o-nedashkivska — APPROVED** [review]

**Review by nikita-remeslov — APPROVED** [review]

**Review by sofiia-trokhymchuk — APPROVED** [review]

**Pull request merged**